File: app/src/main/java/mega/privacy/android/app/modalbottomsheet/ManageTransferBottomSheetDialogFragment.java
package mega.privacy.android.app.modalbottomsheet;

import android.graphics.Bitmap;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import mega.privacy.android.app.AndroidCompletedTransfer;
import mega.privacy.android.app.MegaApplication;
import mega.privacy.android.app.R;
import mega.privacy.android.app.databinding.BottomSheetManageTransferBinding;
import mega.privacy.android.app.main.ManagerActivity;
import mega.privacy.android.app.utils.ColorUtils;
import mega.privacy.android.app.utils.StringResourcesUtils;
import nz.mega.sdk.MegaNode;

import static mega.privacy.android.app.modalbottomsheet.ModalBottomSheetUtil.setThumbnail;
import static mega.privacy.android.app.utils.Constants.INVALID_ID;
import static mega.privacy.android.app.utils.Constants.SNACKBAR_TYPE;
import static mega.privacy.android.app.utils.ThumbnailUtils.*;
import static mega.privacy.android.app.utils.Util.*;
import static nz.mega.sdk.MegaChatApiJava.MEGACHAT_INVALID_HANDLE;
import static nz.mega.sdk.MegaTransfer.*;

public class ManageTransferBottomSheetDialogFragment extends BaseBottomSheetDialogFragment implements View.OnClickListener {
    private static final String TRANSFER_ID = "TRANSFER_ID";
    private static final int MARGIN_TRANSFER_TYPE_ICON_WITH_THUMBNAIL = -12;

    private ManagerActivity managerActivity;

    private AndroidCompletedTransfer transfer;
    private long handle;
    private long transferId;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        BottomSheetManageTransferBinding binding = BottomSheetManageTransferBinding.inflate(getLayoutInflater());
        contentView = binding.getRoot();
        itemsLayout = contentView.findViewById(R.id.item_list_bottom_sheet_contact_file);

        managerActivity = (ManagerActivity) requireActivity();

        if (savedInstanceState == null) {
            transfer = managerActivity.getSelectedTransfer();
            transferId = transfer.getId();
        } else {
            transferId = savedInstanceState.getLong(TRANSFER_ID, INVALID_ID);
            transfer = dbH.getcompletedTransfer(transferId);
        }

        return contentView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        if (transfer == null) return;

        managerActivity = (ManagerActivity) getActivity();

        ImageView thumbnail = contentView.findViewById(R.id.manage_transfer_thumbnail);
        ImageView type = contentView.findViewById(R.id.manage_transfer_small_icon);
        ImageView stateIcon = contentView.findViewById(R.id.manage_transfer_completed_image);
        TextView name = contentView.findViewById(R.id.manage_transfer_filename);
        TextView location = contentView.findViewById(R.id.manage_transfer_location);

        LinearLayout viewInFolderOption = contentView.findViewById(R.id.option_view_layout);
        viewInFolderOption.setOnClickListener(this);
        LinearLayout getLinkOption = contentView.findViewById(R.id.option_get_link_layout);
        getLinkOption.setOnClickListener(this);
        LinearLayout clearOption = contentView.findViewById(R.id.option_clear_layout);
        clearOption.setOnClickListener(this);
        LinearLayout retryOption = contentView.findViewById(R.id.option_retry_layout);
        retryOption.setOnClickListener(this);

        name.setText(transfer.getFileName());

        if (transfer.getType() == TYPE_DOWNLOAD) {
            type.setImageResource(R.drawable.ic_download_transfers);
            getLinkOption.setVisibility(View.GONE);
        } else if (transfer.getType() == TYPE_UPLOAD) {
            type.setImageResource(R.drawable.ic_upload_transfers);
        }

        location.setTextColor(ContextCompat.getColor(requireContext(), R.color.grey_054_white_054));
        RelativeLayout.LayoutParams params =  (RelativeLayout.LayoutParams) stateIcon.getLayoutParams();
        params.rightMargin = dp2px(5, getResources().getDisplayMetrics());

        switch (transfer.getState()) {
            case STATE_COMPLETED:
                location.setText(transfer.getPath());
                stateIcon.setColorFilter(ContextCompat.getColor(requireContext(), R.color.green_500_300), PorterDuff.Mode.SRC_IN);
                stateIcon.setImageResource(R.drawable.ic_transfers_completed);
                retryOption.setVisibility(View.GONE);
                break;

            case STATE_FAILED:
                location.setTextColor(ColorUtils.getThemeColor(requireContext(), R.attr.colorError));
                location.setText(String.format("%s: %s", StringResourcesUtils.getString(R.string.failed_label), transfer.getError()));
                params.rightMargin = 0;
                stateIcon.setImageBitmap(null);
                viewInFolderOption.setVisibility(View.GONE);
                getLinkOption.setVisibility(View.GONE);
                break;

            case STATE_CANCELLED:
                location.setText(R.string.transfer_cancelled);
                params.rightMargin = 0;
                stateIcon.setImageBitmap(null);
                viewInFolderOption.setVisibility(View.GONE);
                getLinkOption.setVisibility(View.GONE);
                break;

            default:
                location.setText(R.string.transfer_unknown);
                stateIcon.clearColorFilter();
                stateIcon.setImageResource(R.drawable.ic_queue);
                break;
        }

        if ((getLinkOption.getVisibility() == View.GONE && retryOption.getVisibility() == View.GONE) || viewInFolderOption.getVisibility() == View.GONE) {
            contentView.findViewById(R.id.separator_get_link).setVisibility(View.GONE);
        }

        stateIcon.setLayoutParams(params);

        handle = Long.parseLong(transfer.getNodeHandle());

        Bitmap thumb = getThumbnailFromCache(handle);
        if (thumb == null) {
            MegaNode node = MegaApplication.getInstance().getMegaApi().getNodeByHandle(handle);
            thumb = getThumbnailFromFolder(node, getContext());
        }

        if (setThumbnail(requireContext(), thumb, thumbnail, transfer.getFileName())) {
            RelativeLayout.LayoutParams typeParams = (RelativeLayout.LayoutParams) type.getLayoutParams();
            typeParams.topMargin = typeParams.rightMargin = MARGIN_TRANSFER_TYPE_ICON_WITH_THUMBNAIL;
            type.setLayoutParams(typeParams);
        }

        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.option_view_layout:
                if (transfer.getType() == TYPE_UPLOAD && !isOnline(requireContext())) {
                    managerActivity.showSnackbar(SNACKBAR_TYPE, getString(R.string.error_server_connection_problem), MEGACHAT_INVALID_HANDLE);
                    break;
                }

                managerActivity.openTransferLocation(transfer);
                break;

            case R.id.option_get_link_layout:
                if (!isOnline(requireContext())) {
                    managerActivity.showSnackbar(SNACKBAR_TYPE, getString(R.string.error_server_connection_problem), MEGACHAT_INVALID_HANDLE);
                    break;
                }

                managerActivity.showGetLinkActivity(handle);
                break;

            case R.id.option_clear_layout:
                managerActivity.removeCompletedTransfer(transfer, true);
                break;

            case R.id.option_retry_layout:
                if (!isOnline(requireContext())) {
                    managerActivity.showSnackbar(SNACKBAR_TYPE, getString(R.string.error_server_connection_problem), MEGACHAT_INVALID_HANDLE);
                    break;
                }

                managerActivity.retrySingleTransfer(transfer);
                break;
        }

        setStateBottomSheetBehaviorHidden();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putLong(TRANSFER_ID, transferId);
    }
}

File: app/src/main/java/mega/privacy/android/app/modalbottomsheet/ManageTransferBottomSheetDialogFragment.kt
package mega.privacy.android.app.modalbottomsheet

import android.content.Intent
import android.graphics.PorterDuff
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.core.content.FileProvider
import androidx.core.net.toUri
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.fragment.app.viewModels
import coil.load
import coil.transform.RoundedCornersTransformation
import mega.privacy.android.app.MimeTypeList
import mega.privacy.android.app.MimeTypeList.Companion.typeForName
import mega.privacy.android.app.R
import mega.privacy.android.app.arch.extensions.collectFlow
import mega.privacy.android.app.databinding.BottomSheetManageTransferBinding
import mega.privacy.android.app.main.ManagerActivity
import mega.privacy.android.app.main.managerSections.CompletedTransfersFragment
import mega.privacy.android.app.utils.ColorUtils
import mega.privacy.android.app.utils.Constants
import mega.privacy.android.app.utils.FileUtil
import mega.privacy.android.app.utils.MegaApiUtils
import mega.privacy.android.app.utils.Util
import mega.privacy.android.app.utils.Util.showSnackbar
import mega.privacy.android.domain.entity.node.NodeId
import mega.privacy.android.domain.entity.node.thumbnail.ThumbnailRequest
import mega.privacy.android.domain.entity.transfer.CompletedTransfer
import mega.privacy.android.domain.entity.transfer.TransferState
import mega.privacy.android.domain.entity.transfer.TransferType
import mega.privacy.android.shared.resources.R as sharedR
import nz.mega.sdk.MegaChatApiJava
import timber.log.Timber
import java.io.File

/**
 * Manage transfer bottom sheet dialog fragment
 *
 */
internal class ManageTransferBottomSheetDialogFragment : BaseBottomSheetDialogFragment(),
    View.OnClickListener {
    private lateinit var managerActivity: ManagerActivity
    private var transfer: CompletedTransfer? = null
    private var parentFileUri: Uri? = null
    private var fileUri: Uri? = null
    private val viewModel by viewModels<ManageTransferSheetViewModel>()
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        val binding = BottomSheetManageTransferBinding.inflate(layoutInflater)
        contentView = binding.root
        itemsLayout = contentView.findViewById(R.id.item_list_bottom_sheet_contact_file)
        managerActivity = requireActivity() as ManagerActivity
        return contentView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewLifecycleOwner.collectFlow(viewModel.uiState) { state ->
            state.transfer?.let {
                transfer = state.transfer
                parentFileUri = state.parentUri
                fileUri = state.fileUri
                handleCompletedTransfer(state)
            }
        }
    }

    private fun handleCompletedTransfer(uiState: ManageTransferSheetUiState) {
        val transfer = transfer ?: run {
            Timber.w("Transfer is not initialized")
            return
        }

        val thumbnail = contentView.findViewById<ImageView>(R.id.manage_transfer_thumbnail)
        val type = contentView.findViewById<ImageView>(R.id.manage_transfer_small_icon)
        val stateIcon = contentView.findViewById<ImageView>(R.id.manage_transfer_completed_image)
        val name = contentView.findViewById<TextView>(R.id.manage_transfer_filename)
        val location = contentView.findViewById<TextView>(R.id.manage_transfer_location)
        val viewInFolderOption = contentView.findViewById<TextView>(R.id.option_view)
        viewInFolderOption.setOnClickListener(this)

        if ((transfer.isContentUriDownload && fileUri != null) || transfer.isContentUriDownload.not()) {
            viewInFolderOption.visibility = View.VISIBLE
        } else {
            viewInFolderOption.visibility = View.GONE
        }

        val openWith = contentView.findViewById<TextView>(R.id.option_open_with)
        openWith.setOnClickListener(this)
        val openWithSeparator = contentView.findViewById<View>(R.id.separator_open_with)
        if (transfer.type == TransferType.DOWNLOAD && fileUri != null) {
            openWith.visibility = View.VISIBLE
            openWithSeparator.visibility =
                if (viewInFolderOption.isGone) View.GONE else View.VISIBLE
        } else {
            openWith.visibility = View.GONE
            openWithSeparator.visibility = View.GONE
        }

        val getLinkOption = contentView.findViewById<TextView>(R.id.option_get_link)
        getLinkOption.text = resources.getQuantityString(
            sharedR.plurals.label_share_links,
            1
        )
        val getLinkOptionSeparator = contentView.findViewById<View>(R.id.separator_get_link)
        getLinkOption.setOnClickListener(this)

        if (uiState.iAmNodeOwner) {
            getLinkOption.visibility = View.VISIBLE
            getLinkOptionSeparator.visibility = View.VISIBLE
        } else {
            getLinkOption.visibility = View.GONE
            getLinkOptionSeparator.visibility = View.GONE
        }

        val clearOption = contentView.findViewById<TextView>(R.id.option_clear)
        clearOption.setOnClickListener(this)
        val retryOption = contentView.findViewById<TextView>(R.id.option_retry)
        val retryOptionSeparator = contentView.findViewById<View>(R.id.separator_retry)
        retryOption.setOnClickListener(this)
        name.text = transfer.fileName
        if (transfer.type.isDownloadType()) {
            type.setImageResource(R.drawable.ic_download_transfers)
        } else if (transfer.type.isUploadType()) {
            type.setImageResource(R.drawable.ic_upload_transfers)
        }
        location.setTextColor(ContextCompat.getColor(requireContext(), R.color.grey_054_white_054))
        when (transfer.state) {
            TransferState.STATE_COMPLETED -> {
                location.text = uiState.parentFilePath ?: transfer.path
                stateIcon.setColorFilter(
                    ContextCompat.getColor(
                        requireContext(),
                        R.color.green_500_300
                    ), PorterDuff.Mode.SRC_IN
                )
                stateIcon.setImageResource(R.drawable.ic_transfers_completed)
                retryOption.visibility = View.GONE
                retryOptionSeparator.visibility = View.GONE
                stateIcon.isVisible = true
            }

            TransferState.STATE_FAILED -> {
                location.setTextColor(
                    ColorUtils.getThemeColor(
                        requireContext(),
                        com.google.android.material.R.attr.colorError
                    )
                )
                location.text = String.format(
                    "%s: %s",
                    getString(R.string.failed_label),
                    transfer.error
                )
                stateIcon.isVisible = false
                viewInFolderOption.visibility = View.GONE
                getLinkOption.visibility = View.GONE
            }

            TransferState.STATE_CANCELLED -> {
                location.setText(R.string.transfer_cancelled)
                stateIcon.isVisible = false
                viewInFolderOption.visibility = View.GONE
                getLinkOption.visibility = View.GONE
            }

            else -> {
                location.setText(R.string.transfer_unknown)
                stateIcon.clearColorFilter()
                stateIcon.setImageResource(R.drawable.ic_queue)
                stateIcon.isVisible = true
            }
        }
        if (getLinkOption.isGone && retryOption.isGone || viewInFolderOption.isGone) {
            getLinkOptionSeparator.visibility = View.GONE
        }
        val thumbParams = thumbnail.layoutParams as FrameLayout.LayoutParams
        thumbnail.load(ThumbnailRequest(NodeId(transfer.handle))) {
            size(Util.dp2px(Constants.THUMB_SIZE_DP.toFloat()))
            transformations(
                RoundedCornersTransformation(
                    Util.dp2px(Constants.THUMB_CORNER_RADIUS_DP).toFloat()
                )
            )
            listener(
                onSuccess = { _, _ ->
                    thumbParams.width = Util.dp2px(Constants.THUMB_SIZE_DP.toFloat())
                    thumbParams.height = thumbParams.width
                    thumbnail.layoutParams = thumbParams
                },
                onError = { _, _ ->
                    thumbParams.width = Util.dp2px(Constants.ICON_SIZE_DP.toFloat())
                    thumbParams.height = thumbParams.width
                    thumbnail.layoutParams = thumbParams
                    thumbnail.setImageResource(MimeTypeList.typeForName(transfer.fileName).iconResourceId)
                },
            )
        }
        calculatePeekHeight()
    }

    override fun onClick(v: View) {
        val transfer = transfer ?: run {
            Timber.w("Transfer is not initialized")
            return
        }

        val id = v.id
        if (id == R.id.option_view) {
            if (transfer.type.isUploadType() && !Util.isOnline(requireContext())) {
                managerActivity.showSnackbar(
                    Constants.SNACKBAR_TYPE,
                    getString(R.string.error_server_connection_problem),
                    MegaChatApiJava.MEGACHAT_INVALID_HANDLE
                )
            } else {
                managerActivity.openTransferLocation(
                    transfer,
                    parentFileUri?.toString() ?: transfer.path
                )
            }
        } else if (id == R.id.option_get_link) {
            if (!Util.isOnline(requireContext())) {
                managerActivity.showSnackbar(
                    Constants.SNACKBAR_TYPE,
                    getString(R.string.error_server_connection_problem),
                    MegaChatApiJava.MEGACHAT_INVALID_HANDLE
                )
            } else {
                managerActivity.showGetLinkActivity(transfer.handle)
            }
        } else if (id == R.id.option_clear) {
            viewModel.completedTransferRemoved(transfer, true)
        } else if (id == R.id.option_retry) {
            if (!Util.isOnline(requireContext())) {
                managerActivity.showSnackbar(
                    Constants.SNACKBAR_TYPE,
                    getString(R.string.error_server_connection_problem),
                    MegaChatApiJava.MEGACHAT_INVALID_HANDLE
                )
            } else {
                (parentFragment as? CompletedTransfersFragment)?.retrySingleTransfer(transfer)
            }
        } else if (id == R.id.option_open_with) {
            openFileWith()
        }
        setStateBottomSheetBehaviorHidden()
    }

    private fun openFileWith() {
        val transfer = transfer ?: run {
            Timber.w("Transfer is not initialized")
            return
        }

        with(transfer) {
            var uri: Uri = originalPath.toUri()

            if (originalPath.startsWith(File.separator)) {
                val localFile = File(originalPath)
                if (FileUtil.isFileAvailable(localFile)) {
                    try {
                        FileProvider.getUriForFile(
                            requireActivity(),
                            Constants.AUTHORITY_STRING_FILE_PROVIDER,
                            localFile
                        )?.let { uri = it }
                    } catch (e: Exception) {
                        Timber.e(e)
                    }
                } else {
                    showSnackbar(requireActivity(), getString(R.string.corrupt_video_dialog_text))
                    return
                }
            } else {
                uri = fileUri ?: uri
            }

            Intent(Intent.ACTION_VIEW).apply {
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                try {
                    setDataAndType(uri, typeForName(fileName).type)
                    val chooserTitle = resources.getString(
                        sharedR.string.open_with_os_dialog_title,
                        fileName
                    )
                    val intent = Intent.createChooser(this, chooserTitle)
                    if (MegaApiUtils.isIntentAvailable(requireActivity(), intent)) {
                        startActivity(intent)
                        return
                    }
                } catch (e: Exception) {
                    Timber.e(e)
                }
            }
        }

        showSnackbar(requireActivity(), getString(R.string.intent_not_available))
    }

    companion object {
        const val TRANSFER_ID = "TRANSFER_ID"
        const val TAG = "ManageTransferBottomSheetDialogFragment"

        fun newInstance(completedTransferId: Int) =
            ManageTransferBottomSheetDialogFragment().apply {
                arguments = Bundle().apply {
                    putInt(TRANSFER_ID, completedTransferId)
                }
            }
    }
}